scrape/scrape.go

```diff
@@ -133,8 +133,79 @@ var (
 			Help: "How many times a scrape cache was flushed due to getting big while scrapes are failing.",
 		},
 	)
+	targetMetadataCache = newMetadataMetricsCollector()
 )
 
+// MetadataMetricsCollector is a Custom Collector for the metadata cache metrics.
+type MetadataMetricsCollector struct {
+	CacheEntries *prometheus.Desc
+	CacheBytes   *prometheus.Desc
+
+	scrapeManager *Manager
+
+	mtx sync.Mutex
+}
+
+func newMetadataMetricsCollector() *MetadataMetricsCollector {
+	return &MetadataMetricsCollector{
+		CacheEntries: prometheus.NewDesc(
+			"prometheus_target_metadata_cache_entries",
+			"Total number of metric metadata entries in the cache",
+			[]string{"scrape_job"},
+			nil,
+		),
+		CacheBytes: prometheus.NewDesc(
+			"prometheus_target_metadata_cache_bytes",
+			"The number of bytes that are currently used for storing metric metadata in the cache",
+			[]string{"scrape_job"},
+			nil,
+		),
+	}
+}
+
+func (mc *MetadataMetricsCollector) registerManager(m *Manager) {
+	mc.scrapeManager = m
+}
+
+// Describe sends the metrics descriptions to the channel.
+func (mc *MetadataMetricsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- mc.CacheEntries
+	ch <- mc.CacheBytes
+}
+
+// Collect creates and sends the metrics for the metadata cache.
+func (mc *MetadataMetricsCollector) Collect(ch chan<- prometheus.Metric) {
+	mc.mtx.Lock()
+	defer mc.mtx.Unlock()
+
+	if mc.scrapeManager == nil {
+		return
+	}
+
+	for _, p := range mc.scrapeManager.scrapePools {
+		var size, length int
```

Comment:
```suggestion
		var size, length float64
```

Reply:
I think `int` is the appropriate choice here. Both `MetadataSize` and `MetadataLen` return `int`, and their sum cannot be larger than the heap size, so the platform dependent `int` size is appropriately sized to not overflow. Conversion to float64 (with precision issues, at least on a 64bit platform) should happen in the end.